typed-racket-lib/typed-racket/typecheck/tc-subst.rkt

```diff
@@ -135,80 +135,48 @@
 (define/cond-contract (subst-object t k o polarity)
   (-> Object? name-ref/c Object? boolean? Object?)
   (match t
-    [(NoObject:) t]
+    [#f t]
     [(Empty:) t]
     [(Path: p i)
      (if (name-ref=? i k)
          (match o
            [(Empty:) -empty-obj]
            ;; the result is not from an annotation, so it isn't a NoObject
-           [(NoObject:) -empty-obj]
+           [#f -empty-obj]
            [(Path: p* i*) (make-Path (append p p*) i*)])
          t)]))
 
-;; Substitution of objects into a filter in a filter set
-;; This is ψ+ [o/x] and ψ- [o/x] with the addition that filters are restricted to
+;; Substitution of objects into a prop in a prop set
+;; This is ψ+ [o/x] and ψ- [o/x] with the addition that props are restricted to
 ;; only those values which are a subtype of the actual argument type (ty).
-(define/cond-contract (subst-filter f k o polarity ty)
-  (-> Filter/c name-ref/c Object? boolean? Type/c Filter/c)
-  (define (ap f) (subst-filter f k o polarity ty))
-  (define (tf-matcher t p i maker)
+(define/cond-contract (subst-prop p k o polarity ty)
+  (-> Prop? name-ref/c Object? boolean? Type/c Prop?)
+  (define (ap q) (subst-prop q k o polarity ty))
+  (define (tprop-matcher pes i t maker)
     (cond
       [(name-ref=? i k)
        (match o
          [(Empty:)
-          (if polarity -top -bot)]
+          (if polarity -tt -ff)]
          [_
           ;; `ty` alone doesn't account for the path, so
           ;; first traverse it with the path to match `t`
-          (define ty/path (path-type p ty))
+          (define ty/path (path-type pes ty))
           (maker
-            ;; don't restrict if the path doesn't match the type
-            (if (equal? ty/path Err)
-                (subst-type t k o polarity ty)
-                (restrict ty/path
-                          (subst-type t k o polarity ty)))
-            (-acc-path p o))])]
-      [(index-free-in? k t) (if polarity -top -bot)]
-      [else f]))
-
-  (match f
-    [(ImpFilter: ant consq)
-     (-imp (subst-filter ant k o (not polarity) ty) (ap consq))]
-    [(AndFilter: fs) (apply -and (map ap fs))]
-    [(OrFilter: fs) (apply -or (map ap fs))]
-    [(Bot:) -bot]
-    [(Top:) -top]
-    [(TypeFilter: t (Path: p i))
-     (tf-matcher t p i -filter)]
-    [(NotTypeFilter: t (Path: p i))
-     (tf-matcher t p i -not-filter)]))
-
-;; Determine if the object k occurs free in the given type
-(define (index-free-in? k type)
```

Comment:
Looking at the paper again (relevant figure is Figure 8, clauses where `x \neq z \and z \in fv(\nu)`), I think I understand the issue better now.

The problem is when `\nu` ends up referring to `x` which is no longer in scope. This means that it could end up in the scope of some _other_ `x`, which would be very bad (or even the same `x`, reinstantiated? maybe?). However, the problem if it ending up in the scope of another `x` can't happen here, because we don't just use symbols for variables. So I think this is ok. But maybe a comment along these lines belongs in the relevant code.